### PR TITLE
[FSDP] relax checking root condition

### PR DESCRIPTION
**Pull request by shuyingsunshine21**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?

When integrating with Lightning, found out that as model is nested in FSDP wrapper after training, and when we call `trainer.test(model)`, it failed the assertion that the root is not set. In this case, non-root has already been set. We relax this assertion in this PR. (link to discussion: https://github.com/PyTorchLightning/pytorch-lightning/pull/6152)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by shuyingsunshine21** [comment]
cc @min-xu-ai , @SeanNaren

**Review by SeanNaren — APPROVED** [review]

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py
Comment (min-xu-ai) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:850]:
```suggestion
                # We relax the assert for non-root instance. A lightning unit test triggers this otherwise.
                assert m._is_root is None or m._is_root == False
```

**Review by min-xu-ai — APPROVED** [review]
Nice. Thanks! Once CI passes I can merge if you don't see the merge button.

**Comment by shuyingsunshine21** [comment]
> Nice. Thanks! Once CI passes I can merge if you don't see the merge button.

looks like CI passes.

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py
Comment (ananthsub) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:850]: not sure if we need to mention lightning here inside of fairscale. eventually this comment will also be unclear what it was relaxed from or why its relaxed

Comment (min-xu-ai) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:850]: I don't want to leave no comment so that it is might be hard to figure out in the future why was this relaxed. Maybe you can suggest something clearer?

**Comment by SeanNaren** [comment]
@ananthsub makes a good point, apologies for being lazy on this one, here is a pure pytorch test we can use to simulate this, and remove the PL line:

```python
import os
import unittest
from unittest import mock

import torch
import torch.nn as nn
from fairscale.nn import FullyShardedDataParallel
import torch.nn.functional as F


@mock.patch.dict(os.environ, {"MASTER_ADDR": "localhost", "MASTER_PORT": "1337"}, clear=True)
@unittest.skipIf(not torch.cuda.is_available(), "Test Requires CUDA")
def test_wrapping_module():
    """
    This test simulates wrapping the module after training to run inference.
    This is required in cases where later in a session, the model is wrapped again in FSDP but
    contains nested FSDP wrappers within the module.
    """
    device = torch.device("cuda")
    torch.cuda.set_device(0)

    torch.distributed.init_process_group(backend="nccl", rank=0, world_size=1)

    module = nn.Sequential(
        FullyShardedDataParallel(nn.Linear(5, 5)),
    )

    model = FullyShardedDataParallel(module).to(device)

    input = torch.rand((1, 5), dtype=torch.float).to(device)
    output = model(input)
    loss = F.mse_loss(input, output)
    loss.backward()

    model = FullyShardedDataParallel(module).to(device)
    second_output = model(input)

    assert torch.allclose(output, second_output)

    torch.distributed.destroy_process_group()
```

**Comment by min-xu-ai** [comment]
@SeanNaren, this is lovely, I can add a test file once my bug 617 work is done.

@shuyingsunshine21, you can add a new test file as well, but please use other fsdp tests as an example. We can't use Sean's code above as is since we don't want to use hard coded tcp port which may cause test port conflict on the same machine when multiple people are running it. Also, a new test file needs to be added to one of the test list text file under tests dir. It is totally fine to leave it to me if you can wait on it a bit.

**Comment by shuyingsunshine21** [comment]
the CI test failure might be related to #624

**Comment by min-xu-ai** [comment]
> 
> 
> the CI test failure might be related to #624

yeah, sorry about that. will be merged within the next hour.

File: tests/ci_test_list_3.txt
Comment (min-xu-ai) [inline tests/ci_test_list_3.txt:8]: can you put the file in list_1.txt since it is shortest right now.

**Comment by shuyingsunshine21** [comment]
> > the CI test failure might be related to #624
> 
> yeah, sorry about that. will be merged within the next hour.

no problem

File: tests/ci_test_list_3.txt
Comment (shuyingsunshine21) [reply tests/ci_test_list_3.txt:8]: was curious about where to put and what is the difference also (so put in similarly place as rest of the fsdp).

**Comment by min-xu-ai** [comment]
Please merge with master. I think Ben has fixed it already. My PR is going in soon too.

File: tests/nn/data_parallel/test_fsdp_multiple_wrapping.py
Comment (min-xu-ai) [inline tests/nn/data_parallel/test_fsdp_multiple_wrapping.py:20]: is this moved by black/isort? If not, CI will fail again. Our CI is pretty strict, it will take a bit of time to get used to. But it is really good once get used to. :-)

**Comment by shuyingsunshine21** [comment]
weird thing why it does not trigger CI

**Comment by min-xu-ai** [comment]
Yeah, I have seen it today too. Perhaps a CI bug. I end up made and pushed a new commit to trigger it.

File: tests/nn/data_parallel/test_fsdp_multiple_wrapping.py
Comment (shuyingsunshine21) [reply tests/nn/data_parallel/test_fsdp_multiple_wrapping.py:17]: thanks, after black forgot to do isort.

**Comment by shuyingsunshine21** [comment]
all passed :)

**Comment by min-xu-ai** [comment]
> 
> 
> all passed :)

Nice! Thanks again!

**Pull request merged**
